src/aleph/netty.clj

```diff
@@ -1543,32 +1551,39 @@
-                          (instance? AddressResolverGroup name-resolver) name-resolver))
-            bootstrap (doto (Bootstrap.)
-                            (.option ChannelOption/SO_REUSEADDR true)
-                            (.option ChannelOption/CONNECT_TIMEOUT_MILLIS (int connect-timeout))
-                            #_(.option ChannelOption/MAX_MESSAGES_PER_READ Integer/MAX_VALUE) ; option deprecated, removed in v5
-                            (.group client-event-loop-group)
-                            (.channel chan-class)
-                            (.handler initializer)
-                            (.resolver resolver')
-                            bootstrap-transform)
-
-            fut (if local-address
-                  (.connect bootstrap remote-address local-address)
-                  (.connect bootstrap remote-address))]
-
-        (d/chain' (wrap-future fut)
-                  (fn [_]
-                    (let [ch (.channel ^ChannelFuture fut)]
-                      (maybe-ssl-handshake-future ch))))))))
+        initializer (pipeline-initializer pipeline-builder)
+        client-event-loop-group @(transport-client-group transport)
+        resolver' (when (some? name-resolver)
+                    (cond
+                      (= :default name-resolver) nil
+                      (= :noop name-resolver) NoopAddressResolverGroup/INSTANCE
+                      (instance? AddressResolverGroup name-resolver) name-resolver))
+        bootstrap (doto (Bootstrap.)
+                        (.option ChannelOption/SO_REUSEADDR true)
+                        (.option ChannelOption/CONNECT_TIMEOUT_MILLIS (int connect-timeout))
+                        #_(.option ChannelOption/MAX_MESSAGES_PER_READ Integer/MAX_VALUE) ; option deprecated, removed in v5
+                        (.group client-event-loop-group)
+                        (.channel chan-class)
+                        (.handler initializer)
+                        (.resolver resolver')
+                        bootstrap-transform)
+
+        fut (connect-client bootstrap remote-address local-address)]
+    (doto (-> (wrap-future fut)
+              (d/chain'
+               (fn [_]
+                 (let [ch (.channel ^ChannelFuture fut)]
+                   (maybe-ssl-handshake-future ch)))))
+      (d/catch' (fn [e]
+                  (when-not (.isDone fut)
+                    (log/trace e "Cancelling Bootstrap#connect future")
+                    (when-not (.cancel fut true)
+                      (when-not (.isDone fut)
+                        (log/warn "Transport" transport "does not support cancellation of connection attempts."
+                                  "Instead, you have to wait for the connect timeout to expire for it to be terminated."
+                                  "Its current value is" connect-timeout "ms."
+                                  "It can be set via the `connect-timeout` option."))))
+                  (d/error-deferred e))))))
```

Comment:
Same issue with the doto and catch, unless I'm misreading the indentation in github. (Also, I don't think `doto` should be indented like this, the `catch'` should be aligned with `(-> (wrap-future ....`

Reply:
> Same issue with the doto and catch, unless I'm misreading the indentation in github.

See https://github.com/clj-commons/aleph/pull/721#discussion_r1561040043

> Also, I don't think `doto` should be indented like this, the `catch'` should be aligned with `(-> (wrap-future ....`

Hmm this is how [`clojure-mode` indents it by default](https://github.com/clojure-emacs/clojure-mode/blob/af0e518a6b86f2c6f32dfb30b99c067071ed5cd4/clojure-mode.el#L1895) and it also corresponds with formatting in `clojure.core`, see e.g. [here](https://github.com/clojure/clojure/blob/527b330045ef35b47a968d80ed3dc4999cfa2623/src/clj/clojure/core.clj#L8022-L8033) :thinking: This is how all forms with a "head" of sorts are commonly indented. Which editor are you using that does it differently? I'd tend to file an issue with them about it.

Reply:
Nm, I'm used to using the arrow fns, which line up the chains and catches. doto is a body macro, and those are technically supposed to be indented only 2 spaces for the body. It just looks out of place, even though it matches the standard style...